cocotb/types/logic.py

```diff
@@ -0,0 +1,279 @@
+# Copyright cocotb contributors
+# Licensed under the Revised BSD License, see LICENSE for details.
+# SPDX-License-Identifier: BSD-3-Clause
+from typing import Any, Optional
+from functools import lru_cache
+
+
+class Logic:
+    r"""
+    Model of a 4-value (``0``, ``1``, ``X``, ``Z``) datatype commonly seen in HDLs.
+
+    .. currentmodule:: cocotb.types
+
+    This is modeled after (System)Verilog's 4-value ``logic`` type.
+    VHDL's 9-value ``std_ulogic`` type maps to this type by treating weak values as full strength values
+    and treating "uninitialized" (``U``) and "don't care" (``-``) as "unknown" (``X``).
+
+    :class:`Logic` can be converted to and from :class:`int`, :class:`str`, :class:`bool`, and :class:`Bit`
+    by using the appropriate constructor syntax.
+    The list of values convertable to :class:`Logic` includes
+    ``0``, ``1``, ``True``, ``False``, ``"0"``, ``"1"``, ``"X"``, ``"Z"``, ``Bit('0')``, and ``Bit('1')``.
+    For a comprehensive list of values that can be converted into :class:`Logic` see :file:`tests/pytest/test_logic.py`.
+
+    .. code-block:: python3
+
+        >>> Logic("X")
+        Logic('X')
+        >>> Logic(True)
+        Logic('1')
+        >>> Logic(1)
+        Logic('1')
+        >>> Logic(Bit(0))
+        Logic('0')
+
+        >>> Logic()  # default value
+        Logic('X')
+
+        >>> str(Logic("Z"))
+        'Z'
+        >>> bool(Logic(0))
+        False
+        >>> int(Logic(1))
+        1
+        >>> Bit(Logic("1"))
+        Bit('1')
+
+    .. note::
+
+        The :class:`int` and :class:`bool` conversions will raise :exc:`ValueError` if the value is not ``0`` or ``1``.
+
+    :class:`Logic` values are hashable and can be placed in :class:`set`\ s and used as keys in :class:`dict`\ s.
+
+    :class:`Logic` supports the common logic operations ``&``, ``|``, ``^``, and ``~``.
+
+    .. code-block:: python3
+
+        >>> def full_adder(a: Logic, b: Logic, carry: Logic) -> (Logic, Logic):
```

Comment:
```suggestion
        >>> def full_adder(a: Logic, b: Logic, carry: Logic) -> Tuple[Logic, Logic]:
```
What you have isn't how annotation are intended to be used